In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending_data = pd.read_csv(
    Path("./Resources/lending_data.csv")
)

# Review the DataFrame
df_lending_data.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending_data['loan_status']

# Separate the X variable, the features
X = df_lending_data.drop(columns=['loan_status'])

In [4]:
# Review the y variable Series
print("Labels (y):")
print(y.head()) 

Labels (y):
0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64


In [5]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [7]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Step 2: Fit the model using the training data
model.fit(X_train, y_train)

# The model is now trained and ready for predictions
print("Logistic Regression model has been successfully trained!")

Logistic Regression model has been successfully trained!


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [8]:
# Make a prediction using the testing data
y_pred = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [9]:
# Generate a confusion matrix for the model
confusion = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(confusion)

Confusion Matrix:
[[14924    77]
 [   31   476]]


In [10]:
# Print the classification report for the model
report = classification_report(y_test, y_pred)
print("\nClassification Report:")
print(report)
print(f'The counts of Healthy Loan Status (0) and Loans with a high likelihood of default (1) are:')

# Show the final counts of the y array to show how many 0s and how many 1s were counted
print(y.value_counts())


Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.86      0.94      0.90       507

    accuracy                           0.99     15508
   macro avg       0.93      0.97      0.95     15508
weighted avg       0.99      0.99      0.99     15508

The counts of Healthy Loan Status (0) and Loans with a high likelihood of default (1) are:
loan_status
0    75036
1     2500
Name: count, dtype: int64


### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Credit Risk Classification Report

# Overview of the Analysis

The purpose of this analysis is to develop a machine learning model capable of predicting loan creditworthiness. Specifically, we aim to classify loans as either "healthy" (label 0) or "high-risk" (label 1) using historical lending activity data from a peer-to-peer lending platform.

Dataset Overview:
Loan Status (Target Variable):

0: Healthy Loan
1: High-Risk Loan

Counts:

Healthy Loans: 75,036
High-Risk Loans: 2,500

Features (Input Variables): Various financial metrics from the dataset that might indicate borrower creditworthiness, excluding the loan_status column.

Machine Learning Process:

Data Preparation:

The dataset was loaded into a Pandas DataFrame, and features (X) and labels (y) were defined.
The data was split into training (80%) and testing (20%) datasets using train_test_split.

Model Training:

A Logistic Regression model was selected due to its suitability for binary classification problems. The model was trained using the training dataset (X_train and y_train).

Model Evaluation:

The model was evaluated using the testing dataset (X_test and y_test).
Performance was assessed via a confusion matrix and classification report, which provided metrics like accuracy, precision, recall, and F1-score.

Results
Logistic Regression Model:
Confusion Matrix:

True Negatives (0 correctly identified as 0): 14,924
False Positives (0 incorrectly identified as 1): 77
False Negatives (1 incorrectly identified as 0): 31
True Positives (1 correctly identified as 1): 476

Performance Metrics:

Accuracy: 99%

Precision:

Class 0 (Healthy Loan): 1.00
Class 1 (High-Risk Loan): 0.86

Recall:

Class 0 (Healthy Loan): 0.99
Class 1 (High-Risk Loan): 0.94

F1-Score:

Class 0 (Healthy Loan): 1.00
Class 1 (High-Risk Loan): 0.90

# Summary

The Logistic Regression model performed exceptionally well in predicting both healthy loans and high-risk loans:

Best Performance: The model excelled in identifying healthy loans (Class 0), achieving near-perfect precision, recall, and F1-scores. This indicates its strength in avoiding false positives for healthy loans.

High-Risk Loans: The model also showed strong recall for high-risk loans (Class 1), accurately identifying 94% of all high-risk cases. However, its precision for this class was slightly lower at 86%, reflecting some false positives.

Recommendations:

Deploy the Model:

Given the high overall accuracy and strong recall for high-risk loans, the model is suitable for deployment, particularly if correctly identifying high-risk loans is a priority.

Future Improvements:

Address the dataset imbalance (only ~3% high-risk loans) to further improve precision for Class 1. Techniques like oversampling or SMOTE could enhance model performance on minority cases.

This Logistic Regression model offers a robust solution for predicting loan creditworthiness and provides actionable insights for financial decision-making.

---